arbos/l1pricing/l1pricing.go

```diff
@@ -208,6 +228,192 @@ func (ps *L1PricingState) UpdateForBatchPosterSpending(
+	}
+	allocationNumerator := updateTime - lastUpdateTime
+	allocationDenominator := currentTime - lastUpdateTime
+	if allocationDenominator == 0 {
+		allocationNumerator = 1
+		allocationDenominator = 1
+	}
+
+	// allocate units to this update
+	unitsSinceUpdate, err := ps.UnitsSinceUpdate()
+	if err != nil {
+		return err
+	}
+	unitsAllocated := unitsSinceUpdate * allocationNumerator / allocationDenominator
+	unitsSinceUpdate -= unitsAllocated
+	if err := ps.SetUnitsSinceUpdate(unitsSinceUpdate); err != nil {
+		return err
+	}
+
+	dueToPoster, err := posterState.FundsDue()
+	if err != nil {
+		return err
+	}
+	err = posterState.SetFundsDue(am.BigAdd(dueToPoster, weiSpent))
+	if err != nil {
+		return err
+	}
+	perUnitReward, err := ps.PerUnitReward()
+	if err != nil {
+		return err
+	}
+	fundsDueForRewards = am.BigAdd(fundsDueForRewards, am.BigMulByUint(am.UintToBig(unitsAllocated), perUnitReward))
+	if err := ps.SetFundsDueForRewards(fundsDueForRewards); err != nil {
+		return err
+	}
+
+	// pay rewards, as much as possible
+	paymentForRewards := am.BigMulByUint(am.UintToBig(perUnitReward), unitsAllocated)
+	availableFunds := statedb.GetBalance(L1PricerFundsPoolAddress)
+	if am.BigLessThan(availableFunds, paymentForRewards) {
+		paymentForRewards = availableFunds
+	}
+	fundsDueForRewards = am.BigSub(fundsDueForRewards, paymentForRewards)
+	if err := ps.SetFundsDueForRewards(fundsDueForRewards); err != nil {
+		return err
+	}
+	payRewardsTo, err := ps.PayRewardsTo()
+	if err != nil {
+		return err
+	}
+	err = util.TransferBalance(
+		&L1PricerFundsPoolAddress, &payRewardsTo, paymentForRewards, evm, scenario, "batchPosterReward",
+	)
+	if err != nil {
+		return err
+	}
+	availableFunds = statedb.GetBalance(L1PricerFundsPoolAddress)
+
+	// settle up our batch poster payments owed, as much as possible
+	allPosterAddrs, err := batchPosterTable.AllPosters(math.MaxUint64)
+	if err != nil {
+		return err
+	}
+	for _, posterAddr := range allPosterAddrs {
+		poster, err := batchPosterTable.OpenPoster(posterAddr, false)
+		if err != nil {
+			return err
+		}
+		balanceDueToPoster, err := poster.FundsDue()
+		if err != nil {
+			return err
+		}
+		balanceToTransfer := balanceDueToPoster
+		if am.BigLessThan(availableFunds, balanceToTransfer) {
+			balanceToTransfer = availableFunds
+		}
+		if balanceToTransfer.Sign() > 0 {
+			addrToPay, err := poster.PayTo()
+			if err != nil {
+				return err
+			}
+			err = util.TransferBalance(
+				&L1PricerFundsPoolAddress, &addrToPay, balanceToTransfer, evm, scenario, "batchPosterRefund",
+			)
+			if err != nil {
+				return err
+			}
+			availableFunds = statedb.GetBalance(L1PricerFundsPoolAddress)
+			balanceDueToPoster = am.BigSub(balanceDueToPoster, balanceToTransfer)
+			err = poster.SetFundsDue(balanceDueToPoster)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// update time
+	if err := ps.SetLastUpdateTime(updateTime); err != nil {
+		return err
+	}
+
+	// adjust the price
+	if unitsAllocated > 0 {
```

Comment:
I feel like surplus is the integral, since price staying at the same level results in the surplus continuously growing. P would be the double derivative, which yeah would require tracking a new thing in state but it may be worth the flexibility

Reply:
I think surplus is the variable we are trying to control, i.e. we are trying to drive the surplus to zero.